### ISSUE TITLE
Breaking changes between version 1.1 and 1.2 (and the future of AutoConfig)

### ISSUE THREAD

**Issue by ben221199** [benbucksch/autoconfig-spec#15]
When comparing my schemas, this are the changes between 1.1 and 1.2:
1. Add `<url>` inside `<incomingServer>`
2. Make `<hostname>`, `<port>`, `<password>` and `<socketType>` inside `<incomingServer>` optional
3. Add `http-basic` as value for `<authentication>` inside `<incomingServer>`
4. Add `jmap`, `ews` and `activesync` to the enum for the `type` attribute on `<incomingServer>`
5. Make `<password>` inside `<outgoingServer>` optional
6. Add `<addressBook>`, `<calendar>`, `<fileShare>`, `<mAuth>`, `<enable>`, `<clientConfigUpdate>` on `<clientConfig>`
7. Add `<inputField>` on `<clientConfig>`

There may be more changes that I missed, but I will go with these for now.

---

By the way that ThunderBird is implemented, I assume that extended enums (change 3 and 4) will break the code. Also, in change 1, `<url>` is used instead of `<hostname>` and `<port>`, which likely breaks many clients that expect an hostname and port. Change 7 seems to me like a mistake, because `<inputField>` was inside `<emailProvider>` before. And last, I don't like change 6.

In my conclusion, we can minimalize the breaking changes problem to introduce a new 1.2 URL instead of building on the same URL. 

---

Also, I don't think we should add all those other types of services in this AutoConfig. Currently, the 1.1 URL is `http://autoconfig.$DOMAIN/mail/config-v1.1.xml?emailaddress=$EMAILADDRESS` (or `http://$DOMAIN/.well-known/autoconfig/mail/config-v1.1.xml?emailaddress=$EMAILADDRESS`). Notice the `mail` part in the URL, so let this be AutoConfig Mail, and lets make an AutoConfig Calendar and AutoConfig Contacts seperately. We may link those URLs in some general AutoConfig file.

For now, XML is fine for 1.2, but when developing 2.0, the switch to JSON seems pretty straight forward.

**Comment by ben221199** [comment]
Example of my idea:

## AutoConfig 1.*

The XML versions.

AutoConfig (new spec):
```xml
<clientConfigList version="1.0">
  <clientConfig url="https://autoconfig.mozzila.com/mail/config-v1.1.xml" type="mail" version="1.1"/>
  <clientConfig url="https://autoconfig.mozzila.com/mail/config-v1.2.xml" type="mail" version="1.2"/>
  <clientConfig url="https://autoconfig.mozzila.com/calendar/config-v1.0.xml" type="calendar" version="1.0"/>
  <clientConfig url="https://autoconfig.mozzila.com/contacts/config-v1.0.xml" type="contacts" version="1.0"/>
</clientConfigList>
```

AutoConfig Mail (<ins>**this spec**</ins>):
```xml
<clientConfig version="1.0, 1.1 or 1.2">
  <emailProvider>
    ...
  </emailProvider>
</clientConfig>
```

AutoConfig Calendar (new other spec):
```xml
<clientConfig version="1.0">
  <calendarProvider>
    ...
  </calendarProvider>
</clientConfig>
```

AutoConfig Contacts (new other spec):
```xml
<clientConfig version="1.0">
  <contactsProvider>
    ...
  </contactsProvider>
</clientConfig>
```

## AutoConfig 2.*'

The JSON versions.

AutoConfig (all specs in one):
```json
{
  "providers" : [
    ...
  ]
}
```

**Comment by benbucksch** [comment]
> When comparing my schemas, this are the changes between 1.1 and 1.2:
>
>     Add <url> inside <incomingServer>
>     Make <hostname>, <port>, <password> and <socketType> inside <incomingServer> optional


Does XSD support something like "subclasses"? That you can define `<incomingServer>` more generally, and then specifically for each protocol?

In 1.2 (the current spec), a config 
```
<incomingServer type="imap">
  <url>https://example.com</url>
</incomingServer>
```
is **not** valid. And a JMAP server without URL and with hostname and port is not valid.

Version 1.1 supported only IMAP and POP3 and therefore the problem didn't appear.

>     Add `http-basic` as value for `<authentication>` inside `<incomingServer>` 


Ditto. Not valid for IMAP, but only for configs with a URL.


>     Make `<password>` inside `<outgoingServer>` optional


Password was always optional for both incomingServer and outgoingServer


>     Add `<addressBook>, <calendar>, <fileShare>, <mAuth>, <enable>, <clientConfigUpdate>`
>     Add `<inputField>`


`<enable>`, `<clientConfigUpdate>`, `<inputField>` were already there in 1.1 spec, but are not used by Thunderbird.

https://wiki.mozilla.org/index.php?title=Thunderbird:Autoconfiguration:ConfigFileFormat&oldid=1227306 (revision from 2020)

**Comment by benbucksch** [comment]
> By the way that ThunderBird is implemented, I assume that extended enums (change 3 and 4) will break the code. Also, in change 1, <url> is used instead of <hostname> and <port>, which likely breaks many clients that expect an hostname and port.

hm, interesting. I checked. From how I wrote the Thunderbird code back then, and how I read it now, there is a [try/catch](https://searchfox.org/comm-central/source/mail/components/accountcreation/modules/readFromXML.sys.mjs#196) around the processing of an `<incomingServer>` element. Yes, `type="jmap"` would throw, it would be cause *inside* the `for` loop, and the code would start reading the next `<incomingServer>`. After the loop, there's a check whether *any* valid `<incomingServer>` was found. If so, that config is used and the "invalid" (actually not yet supported) config is simply ignored. If no config at all is found (`if (!d.incoming.hostname)`), then it would throw the error and stop.

So, effect is: Thunderbird would ignore an `<incomingServer type="jmap">` as if it doesn't exist, and continue reading the rest of the config. So, I don't see a problem.

Would you concur with my reading of the code?

You have a good point, though: The spec should mandate clients to ignore config parts that the client cannot understand, but process the rest. This is the only way that we can continue evolving the spec. Otherwise we're completely stuck.

**Comment by benbucksch** [comment]
> Example of my idea:
> ```
> <clientConfigList version="1.0">
>   <clientConfig url="https://autoconfig.mozzila.com/mail/config-v1.1.xml" type="mail" version="1.1"/>
>   <clientConfig url="https://autoconfig.mozzila.com/mail/config-v1.2.xml" type="mail" version="1.2"/>
>   <clientConfig url="https://autoconfig.mozzila.com/calendar/config-v1.0.xml" type="calendar" version="1.0"/>
>   <clientConfig url="https://autoconfig.mozzila.com/contacts/config-v1.0.xml" type="contacts" version="1.0"/>
> </clientConfigList>
> ```

Thanks for the idea. But I fundamentally disagree.

1. All services that a hoster typically provides to a user and that mail apps may implement should be in there. Even chat and video conference accounts. The client should have to read only this one file to set up this email account with all associated services.

2. Creating multiple protocols or URLs creates more burden for implementors, and we'll need to make even more URL queries. We already have many, to cover different scenarios and providers. I don't want to multiply that by 4 or 6. You also increase possibilities to fail.

3. Another way to look at this is: An email client should also sync contacts. If it does calendar, and the email account supports calendar sync, why would you not have that set up? If you do, why not do it in the same doc?

4. The separation makes even less sense for protocols like JMAP, EWS, ActiveSync, MS Graph. Basically *all* modern mail protocols come with contact and calendar sync already built in, with the same connection. Why artifically separate it fpr autoconfig, when it's already united in the server protocols?

**Comment by ben221199** [comment]
> You have a good point, though: The spec should mandate clients to ignore config parts that the client cannot understand, but process the rest. This is the only way that we can continue evolving the spec. Otherwise we're completely stuck.

I have no problems with the current spec being strict. We are talking about XML here, so having a XML namespace with some rules seems fine with me.

I think we can drop this strictness as soon as we start working on AutoConfig 2.0 with JSON, but for XML I think we need some form of strictness.

**Comment by ben221199** [comment]
> > Example of my idea:
> > ```
> > <clientConfigList version="1.0">
> >   <clientConfig url="https://autoconfig.mozzila.com/mail/config-v1.1.xml" type="mail" version="1.1"/>
> >   <clientConfig url="https://autoconfig.mozzila.com/mail/config-v1.2.xml" type="mail" version="1.2"/>
> >   <clientConfig url="https://autoconfig.mozzila.com/calendar/config-v1.0.xml" type="calendar" version="1.0"/>
> >   <clientConfig url="https://autoconfig.mozzila.com/contacts/config-v1.0.xml" type="contacts" version="1.0"/>
> > </clientConfigList>
> > ```
> 
> Thanks for the idea. But I disagree here.
> 
> 1. All services that hosters provide to users and that mail apps may reasonably implement should be in the config. Even chat and video conference accounts. The mail client should have to read only this one file to set up this email account with all associated services.

I agree that having to check as less files as possible is the best approach, yes.

> 2. Creating multiple separate autoconfig protocols or URLs creates more burden for implementors, and we'll need to make even more URL queries. We already have many URL queries, to cover different scenarios and providers. I don't want to multiply by 4 or 6. It would also increase the possibilities to fail.

Agreed, but sometimes the history of a spec forces you to introduce a new URL, in this case because this has been done before with 1.0 to 1.1 and because there could be breaking changes.

> 3. An email client should also sync contacts, otherwise it's hard to send mail. If it does calendar, and the email account supports calendar sync, why would you not have that set up? If you do, why not do it in the same doc?

Yeah, like how the ActiveSync is a protocol for all those things. I understand. However, the current URL has `/mail/` in it, so that is why I call it `AutoConfig Mail` and not just `AutoConfig`. Is there a reason to keep the word `/mail/` in it, except for backward compatibility and historic stuff?

> 4. The separation makes even less sense for protocols like JMAP, EWS, ActiveSync, MS Graph. Basically _all_ modern mail protocols come with contact and calendar sync already built in, on the very same server connection. Why artifically separate it for autoconfig, when it's already united in the server protocols?

I made the seperation, because it may be an option to want a calendar (ICS/CalDAV) or a contact list (VCF/CardDAV) without any mail, where all those for protocols/formats are also distinct things (like IMAP/POP and SMTP).

---

I call the protocol `AutoConfig Mail` (instead of just `AutoConfig`) for a reason. This is because the URL contains the part `/mail/`, but also because is mainly focussed on mail. In my head it a bit strange to have `AutoConfig Mail` also configure your calendar, contacts and even cloud space. There may be reasons for it to keep it `AutoConfig Mail` (with all those configurations), but if you are thinking about dropping the `Mail` part (which seems more logical in that case), an URL without `/mail/` should be used and the versioning should be `1.0` again (because we are technically talking about a new format). This means that, from that moment, you are free to design a fully a new format. This can be XML again, or you could directly choose to do JSON.

**Comment by benbucksch** [comment]
> Is there a reason to keep the word /mail/ in it, except for backward compatibility and historic stuff?

No, purely for backward compat with existing servers and clients. Mostly the servers. Otherwise we'd also add `.well-known/`. But that's just a string, just ignore what it says.

**Comment by benbucksch** [comment]
> You have a good point, though: The spec should mandate clients to ignore config parts that the client cannot understand, but process the rest. This is the only way that we can continue evolving the spec. Otherwise we're completely stuck.

I've added explicit text that the client MUST ignore elements, properties and type/auth values that it doesn't understand.

The latest revision of the spec text is a major overhaul. Please have another read.

FIXED

**Comment by ben221199** [comment]
Okay, nice. I will take a look at this repository soon, because I haven't yet made/finished a XSD for 1.2 yet.

**Comment by benbucksch** [comment]
Current draft status is here: https://www.ietf.org/archive/id/draft-ietf-mailmaint-autoconfig-01.html
